### PR TITLE
First version?

### PR DESCRIPTION
**Pull request by aviramha**

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+        .args(&["-c", "import sys; print(sys.version_info[1])"])
+        .output()
+        .expect("python version did not print");
+    let minor = u8::from_str_radix(String::from_utf8_lossy(&out.stdout).trim(), 10)
+        .expect("python version was not parsed");
+
+    for i in 6..=minor {
+        println!("cargo:rustc-cfg=Py_3_{}", i);
+    }
+    println!("cargo:rustc-cfg=Py_3");
```
Comment (davidhewitt) [inline build.rs:14]: Yeah I'd like to make a pyo3-config crate sometime so that downstream projects can re-use the pyo3 logic :)

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+            _ => return 0,
+        },
+        _ => return 0,
+    };
+
+    let frame_data = match FrameData::try_from(frame) {
+        Ok(v) => v,
+        Err(err) => {
+            println!("{}", err);
+            return 0;
```
Comment (davidhewitt) [inline src/lib.rs:179]: For error handling you might be able to use `PyErr::restore` in combination with `return -1` to raise exceptions.

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+}
+
+fn get_context(py: Python, obj: *mut pyo3::ffi::PyObject) -> Option<Py<ProfilerContext>> {
+    let mut context_value: *mut pyo3::ffi::PyObject = std::ptr::null_mut();
+
+    unsafe {
+        match PyContextVar_Get(
+            obj,
+            std::ptr::null_mut(),
+            &mut context_value as *mut *mut pyo3::ffi::PyObject,
```
Comment (davidhewitt) [inline src/lib.rs]: I think `&mut context_value` should be enough here.

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        ) {
+            0 => (),
+            _ => return None,
```
Comment (davidhewitt) [inline src/lib.rs:149]: I think you might want to handle errors here, but it'll mean changing the function to return `PyResult<Option<...>>:

```suggestion
            -1 => return Err(PyErr::fetch()),
            _ => return None,
```

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    let context_obj =
+        match unsafe { Py::<ProfilerContext>::from_borrowed_ptr_or_opt(py, context_value) } {
+            Some(v) => v,
+            None => return None,
+        };
+
+    unsafe {
+        pyo3::ffi::Py_XDECREF(context_value);
+    }
+    Some(context_obj)
```
Comment (davidhewitt) [inline src/lib.rs]: `from_owned_ptr_or_opt` handles the null check and the decref.

```suggestion
    unsafe { Py::from_owned_ptr_or_opt(py, context_obj) }
```

**Review by davidhewitt — COMMENTED** [review]
I took a quick read of the pyo3 stuff - could be interesting. Most of my ideas are around error handling.

File: src/lib.rs (hunk 4 above)
Comment (aviramha) [reply src/lib.rs:149]: This function is from a callback, so I can't really raise Python errors I believe

File: build.rs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+use std::process::Command;
+
+fn main() {
+    let out = Command::new("python")
```
Comment (guyzyl) [inline build.rs:4]: Isn't there a better way to check if Python is/n't installed?

File: capara/profiler.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import sys
+from contextvars import ContextVar
+from typing import List, Optional, Tuple
+
+from . import capara
+
+_reference_count = 0
```
Comment (guyzyl) [inline capara/profiler.py:8]: Why not use a class instead `global _reference_count`?

File: capara/profiler.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    if _reference_count == 0:
+        capara.start(_profiler_context)
+    _reference_count += 1
+
+
+def stop() -> List[Tuple[str, str, Optional[int]]]:
+    """Stops the profiler. Completely stops the profiler only if reference count equals to zero.
+
+    Returns:
+        List of profiler events, each event is a tuple of (file_name, func_name, duration).
```
Comment (guyzyl) [inline capara/profiler.py]: Working with tuples suck (in my opinion), consider using a `dataclass` (something like `ProfileResult`), will make the code nicer

File: tests/requirements.txt [hunk 9]
```diff
@@ -0,0 +1 @@
+pytest>=6.1.2
```
Comment (guyzyl) [inline tests/requirements.txt]: Why is a `requirements.txt` file needed if you're using `Poetry`? Also, should be `requirements.dev.txt`

**Review by guyzyl — COMMENTED** [review]
Wrote some comments

File: capara/profiler.py (hunk 7 above)
Comment (guyzyl) [reply capara/profiler.py:8]: Maybe move all operations against `_reference_count` to an external function? Something like `update_global_reference_count`

File: src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+            entries: std::cell::Cell::new(HashMap::new()),
+        }
+    }
+    #[getter]
+    fn entries(&mut self) -> PyResult<Vec<(String, String, Option<u128>)>> {
+        let mut result = Vec::new();
+        for (_, entry) in self.entries.get_mut().iter() {
+            result.push(format_entry(entry));
+        }
+        Ok(result)
```
Comment (oskardrums) [inline src/lib.rs]: Another nit:
```suggestion
    Ok(self.entries.get_mut().iter().map(|_, entry| format_entry(entry)))
```

File: src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+struct ProfilerContext {
+    entries: std::cell::Cell<HashMap<usize, ProfilerEntry>>,
+}
+
+/// Format an entry into file_name, func_name and duration.
+fn format_entry(entry: &ProfilerEntry) -> (String, String, Option<u128>) {
+    let duration = match entry.end {
+        Some(v) => Some(v.duration_since(entry.start).as_nanos()),
+        None => None,
+    };
```
Comment (oskardrums) [inline src/lib.rs]: nit, use combinators:
```suggestion
    let duration = match entry.end.and_then(|v| ...);
```

File: src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+                && (!dframe.f_stacktop.is_null())
+        };
+
+        unsafe {
+            let file_name = match code.co_filename.is_null() {
+                true => "null".to_string(),
+                false => CStr::from_ptr(PyUnicode_AsUTF8(code.co_filename))
+                    .to_string_lossy()
+                    .into_owned(),
+            };
```
Comment (oskardrums) [inline src/lib.rs]: Consider extracting this bit into a function. It's repeated in the following line so it would do good for code reuse as well as better constraining the `unsafe` code

File: src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+        if dframe.f_code.is_null() {
+            return Err("f_code is null");
+        }
+
+        let code = unsafe { *dframe.f_code };
+        let yielded_coroutine = {
+            (0 < (code.co_flags & CO_COROUTINE)
+                | (code.co_flags & CO_ITERABLE_COROUTINE)
+                | (code.co_flags & CO_ASYNC_GENERATOR))
+                && (!dframe.f_stacktop.is_null())
```
Comment (oskardrums) [inline src/lib.rs]: nit:
```suggestion
let yielded_coroutine = (code.co_flags & (CO_COROUTINE | CO_ITERABLE | ...) > 0) && ...
```

File: src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+    }
+}
+
+/// Struct representing frame data
+/// Identifier is frame pointer casted to usize.
+struct FrameData {
+    func_name: String,
+    file_name: String,
+    identifier: usize,
+    yielded_coroutine: bool,
```
Comment (oskardrums) [inline src/lib.rs]: consider `is_yielded_coroutine` or `did_yield_coroutine`, English is ambiguous otherwise :\

File: src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+impl TryFrom<c_int> for TraceEvent {
+    type Error = &'static str;
+    /// Cast i32 event (raw from Python) to Rust enum.
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        if value > 7 || value < 0 {
+            return Err("Not valid enum value");
+        }
+        Ok(unsafe { std::mem::transmute(value) })
```
Comment (oskardrums) [inline src/lib.rs]: Don't hardcode the number of `TraceEvent` variants, instead maybe have
```suggestion
const LARGER_THAN_ANY_TRACE_EVENT: i32 = TraceEvent::Opcode as i32 + 1;
...
        if value >= LARGER_THAN_ANY_TRACE_EVENT {...}
```

Another approach would be matching `value` against every variant's numeric value. That way you don't need the `unsafe` call to `transmute`, and because `TraceEvent` is non-exhaustive whenever you add a new variant it would have to be matched against to get the code to compile, which is also good. You might pay a little in performance, though.

File: src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+enum TraceEvent {
+    Call,
+    Exception,
+    Line,
+    Return,
+    CCall,
+    CException,
+    CReturn,
+    Opcode,
+    __NonExhaustive,
```
Comment (oskardrums) [inline src/lib.rs]: `#[non_exhaustive]` is stable, you can use it on `TraceEvent` instead of the `__NonExhaustive` variant trick

File: src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+    };
+
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+
+    let context = match get_context(py, obj) {
+        Some(v) => v,
+        _ => return 0,
+    };
+    let mut context = context.as_ref(py).borrow_mut();
```
Comment (oskardrums) [inline src/lib.rs:190]: Consider something like:
```suggestion
    let mut context = get_context(py, obj)
        .and_then(|v| Some(v.as_ref(py).borrow_mut()))
        .or(0);
```

File: src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+        }
+        TraceEvent::Return => {
+            if frame_data.yielded_coroutine {
+                return 0;
+            }
+            if let Some(entry) = context.entries.get_mut().get_mut(&frame_data.identifier) {
+                entry.end = Some(Instant::now());
+            };
+        }
+        _ => println!("shouldn't happen"),
```
Comment (oskardrums) [inline src/lib.rs]:
```suggestion
        _ => unreachable!(),
```

**Review by oskardrums — COMMENTED** [review]
Hey,
I went over the rust parts and added some comments, most of them are nits, all in all looks really good :)

File: build.rs (hunk 6 above)
Comment (aviramha) [reply build.rs:4]: Copied from other PyO3.. I didn't find any better way.

File: src/lib.rs (hunk 17 above)
Comment (aviramha) [reply src/lib.rs:190]: It didn't seem to work - I think or(0) doesn't return 0 in the main function.. which is what I need :(

Comment (oskardrums) [reply src/lib.rs:190]: Oh right, `or` takes (and returns) an `Option`, so here you would need `unwrap_or(0)` instead.
Actually going back over [`Option`'s docs](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.map_or) I think it can also be written as
```suggestion
    let mut context = get_context(py, obj).map_or(0, |v| v.as_ref(py).borrow_mut());
```
that's quite concise :)

File: capara/profiler.py [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+        capara.start(_profiler_context)
+    _reference_count.count += 1
+
+
+def stop() -> ProfilerResult:
+    """Stops the profiler. Completely stops the profiler only if reference count equals to zero.
+
+    Returns:
+        ProfilerResult
+    """
```
Comment (bharel) [inline capara/profiler.py:46]: Bug: Check if ref is 0 beforehand. Otherwise it'll progress to -1.

File: capara/profiler.py [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+
+@dataclass
+class ProfilerEntry:
+    func_name: str
+    file_name: str
+    duration: Optional[int] = None
+
+    @classmethod
+    def from_tuple(cls, file_name: str, func_name: str, duration: Optional[int]) -> "ProfilerEntry":
+        return cls(file_name=file_name, func_name=func_name, duration=duration)
```
Comment (bharel) [inline capara/profiler.py]: Why not use NamedTuple?

File: capara/profiler.py [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+from contextvars import ContextVar
+from dataclasses import dataclass
+from typing import List, Optional
+
+from . import capara
+
+
+@dataclass
+class ReferenceCount:
+    count: int = 0
```
Comment (bharel) [inline capara/profiler.py]: Why did you change it? It was fine as a global int.

File: capara/profiler.py [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+    """Checks if profiler is active for current context."""
+    return _profiler_context.get() is not None
+
+
+class Profiler:
+    def __init__(self):
+        self.results: Optional[ProfilerResult] = None
+
+    def __enter__(self):
+        start()
```
Comment (bharel) [inline capara/profiler.py:72]:
```suggestion
    def __enter__(self):
        start()
        return self
```

File: capara/profiler.py [hunk 23]
```diff
@@ -0,0 +1,82 @@
+import sys
```
Comment (bharel) [inline capara/profiler.py:1]: Add a note somewhere you do not support multi-threading.

File: src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,221 @@
...
+            };
+            context
+                .entries
+                .get_mut()
+                .insert(frame_data.identifier, entry);
+        }
+        TraceEvent::Return => {
+            if frame_data.is_yielded_coroutine {
+                return 0;
+            }
```
Comment (bharel) [inline src/lib.rs:223]: !!! There is an issue that pertains to your entire logic. It's the same one about yield that I warned you before. You calculate cumulative time (cumtime) everywhere. Usually you'd want total time (tottime) in asyncio scenarios, as cumtime will include time spent processing other non-related CPU intensive code. !!!  (I wouldn't even call it cumtime)

Your code might be perfect per-se, but right now the profiler is unfortunately useless.

You can add a new field for tottime and increase it every yield. It will require that you'll track every call though.

File: src/lib.rs [hunk 25]
```diff
@@ -0,0 +1,221 @@
...
+                func_name: frame_data.func_name,
+                file_name: frame_data.file_name,
+                start,
+                end: None,
+            };
+            context
+                .entries
+                .get_mut()
+                .insert(frame_data.identifier, entry);
+        }
```
Comment (bharel) [inline src/lib.rs]: When you continue from a `yield`, is a `Call` event thrown? If the answer is yes, you don't handle this. If the answer is no, you have an even bigger issue over here.

**Review by bharel — CHANGES_REQUESTED** [review]

**Comment by bharel** [comment]
To further elaborate my review, you can test the cumtime/whatever issue like so:

```
async def test():
 time.sleep(5)
asnyc def test2():
 await asyncio.sleep(1)
```

If you profile both `test()`, and `test2()` together, `test2()` will take 5 seconds, which is just plain wrong.

See, it's not even cumtime. Your profiler includes random unrelated variables. As long as you don't account for it, it's kinda useless.

File: capara/profiler.py [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+from . import capara
+
+_reference_count = 0
+_profiler_context: ContextVar[Optional[capara.ProfilerContext]] = ContextVar("profiler_context", default=None)
+
+
+class ProfilerEntry(NamedTuple):
+    file_name: str
+    func_name: str
+    duration: Optional[int]
```
Comment (bentheiii) [inline capara/profiler.py:16]: shouldn't it be a float? if not, the unit should be specified

File: src/lib.rs [hunk 27]
```diff
@@ -0,0 +1,250 @@
...
+struct ProfilerContext {
+    entries: Cell<HashMap<usize, ProfilerEntry>>,
+    count: i32,
+}
+
+/// Format an entry into file_name, func_name and duration.
+fn format_entry(entry: &ProfilerEntry) -> (String, String, Option<u128>, i32) {
+    let duration = entry
+        .end
+        .and_then(|v| Some(v.duration_since(entry.start).as_nanos()));
```
Comment (bentheiii) [inline src/lib.rs:114]: note that i32 in nanos can store just over 2 seconds

File: src/lib.rs [hunk 28]
```diff
@@ -0,0 +1,250 @@
...
+fn start(context_var: &PyAny) -> PyResult<()> {
+    let cb: Py_tracefunc = callback;
+    unsafe {
+        PyEval_SetProfile(cb, context_var.as_ptr());
+    }
+    Ok(())
+}
+
+#[pymodule]
+/// A Python module implemented in Rust.
```
Comment (bentheiii) [inline src/lib.rs]: remove
